File: Solutions/98-Validate Binary Search Tree/98-Validate Binary Search Tree.py
#Fourth fastest
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        def check(root, minimum, maximum):
            if not root:
                return True
            
            if root.val <= minimum or root.val >= maximum:
                return False
            
            left, right = True, True 
            if root.left:
                left = check(root.left, minimum, root.val)
            if root.right:
                right = check(root.right, root.val, maximum)
            
            return left and right
        return check(root, float("-inf"), float("inf"))

#Third fastest
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def rec(root, a, b):
            return not root or (a < root.val < b and rec(root.left,a, root.val) and rec(root.right, root.val, b))
        
        return rec(root, float("-inf"), float("inf"))

#Second fastest
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def isValid(root, mini, maxi):
            if not root:
                return True
            if root.val >= maxi or root.val <= mini:
                return False
            return isValid(root.left, mini, root.val) and isValid(root.right, root.val, maxi)
            
            
        return isValid(root, float('-inf'), float('inf'))

#Fastest
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        if not root:
            return True
        stack = []
        pre = None
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if pre != None and pre.val >= root.val:
                return False
            pre = root
            root = root.right
            
        return True
            
                
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
#         if not root:
#             return True
            
#         stack = [(float('-inf'), root, float('inf'))] 
#         while stack:
#             lower, root, upper = stack.pop()
#             if not root:
#                 continue
#             val = root.val
#             if val <= lower or val >= upper:
#                 return False
#             stack.append((val, root.right, upper))
#             stack.append((lower, root.left, val))
#         return True  

File: Solutions/98-Validate Binary Search Tree/98-Validate Binary Search Tree.cpp
//Third fastest
class Solution {
public:
	queue<int> q;
	int pass = 1;

	void DFS(TreeNode* root) {

		if(pass != 1)
			return;

		if(root != NULL){
			DFS(root->left);

			if(q.empty())
				q.push(root->val);
			else{
				if(root->val > q.front()){
					q.pop();
					q.push(root->val);
				}
				else
					pass = 0;
			}
			DFS(root->right);
		}
		return;
	}

    bool isValidBST(TreeNode* root) {
		int tmp1;
        int tmp2;
		DFS(root);

		if(pass == 1)
			return true;
		return false;
    }
};

//Second fastest
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(root==nullptr||(root->left==nullptr&&root->right==nullptr)) return true;
        if(root->left!=nullptr &&root->left->val>=root->val) return false;
        if(root->right!=nullptr &&root->right->val<=root->val) return false;
        bool isvalidL= isValidBST(root->left);
        if((root->left!=nullptr&&root->left->val>=root->val)||(root->left!=nullptr&&maxleft(root->left)>=root->val)) return false;
        bool isvalidR= isValidBST(root->right);
        if((root->right!=nullptr&&root->right->val<=root->val)||(root->right!=nullptr&&minright(root->right)<=root->val)) return false;
        return isvalidL&isvalidR;
    }
    
    int maxleft(TreeNode* left){
        while(left->right!=nullptr) left = left->right;
        return left->val;
    }
    
    int minright(TreeNode *right){
        while(right->left!=nullptr) right = right->left;
        return right->val;
    }
};

//Fastest
class Solution {
private:
    TreeNode* prev=NULL;
public:
    bool isValidBST(TreeNode* root) {
        if(!root)   return true;
        return isValidBST(root->left)  && help(root) && isValidBST(root->right);
    }
    
    bool help(TreeNode* root){
        if(!prev){
            prev=root;
            return true;
        }
        if(prev->val >= root->val)   return false;
        prev=root;
        return true;
    }
};